SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/ReferenceCountedLockingPolicy.cpp

```diff
@@ -127,3 +132,28 @@ void UReferenceCountedLockingPolicy::OnLockedActorDeleted(AActor* DestroyedActor
 	}
 	ActorToLockingState.Remove(DestroyedActor);
 }
+
+bool UReferenceCountedLockingPolicy::AcquireLockFromDelegate(AActor* ActorToLock, const FString& DelegateLockIdentifier)
+{
+	ActorLockToken LockToken = AcquireLock(ActorToLock, DelegateLockIdentifier);
+	if (LockToken == SpatialConstants::INVALID_ACTOR_LOCK_TOKEN)
+	{
+		UE_LOG(LogReferenceCountedLockingPolicy, Error, TEXT("AcquireLock called from engine delegate returned an invalid token"));
+		return false;
+	}
+
+	check(!DelegateLockingIdentifierToActorLockToken.Contains(DelegateLockIdentifier));
+	DelegateLockingIdentifierToActorLockToken.Add(DelegateLockIdentifier, LockToken);
```

Comment:
Is passing the ActorLockToken to the calling code a no-go ?
The exposed API makes the String mandatory, while the locking policy API had it as an optional debug string.

Reply:
> Is passing the ActorLockToken to the calling code a no-go ?

Do you mean, why don't we return the ActorLockToken in the delegate? If so, the reason is being it's not necessary, the ability activation can provide its own uniqueness in terms of handles that are embedded into the LockIdentifier string. ActorLockTokens are useful for providing uniqueness to a system requesting lock that might not be in this position. We could ignore the LockIdentifier being sent from engine code at all, but that requires adding more state to ability activations and engine code which we'd preferably avoid imo.

Reply:
Maybe the Gameplay ability system is able to provide its own identifier, but the way the GDK service is used through these interfaces is adding additional unecessary constraints. The original API looked like a traditional Mutex API, easy to use and understand, something that could be adapted to other systems having different properties compared to the GAS. 
This API is making a choice for every engine user, that using the GDK locks HAS to make use of a FString that HAS to be unique accross all calls, something that was not required from the original API, which was simpler and more efficient (no string allocation, no requirement for an external identifier allocation).

Reply:
I would have less issues with a FString -> Token map being hosted in the GAS side, as opposed to the GDK side.

Reply:
> the way the GDK service is used through these interfaces is adding additional unecessary constraints

Not sure I understand this? Can you elaborate

> something that could be adapted to other systems having different properties compared to the GAS

Providing FString lock identifiers was chosen as something that any engine system could embed its own uniqueness within. There are advantages of giving the GAS system the ability to do this that I mentioned above such as not adding more state to ability activations. There may well be a desire for another engine system to acquire a lock without providing an identifier, but implementing those delegates with no requirement for it at this stage seems overly preemptive. 

> I would have less issues with a FString -> Token map being hosted in the GAS side, as opposed to the GDK side.

I can understand and did consider it. It's a tradeoff of effectively overloading functions in the locking interface but at the benefit of being able to include them in testing.

Reply:
So this is essentially the difference between the caller being forced to generate a deterministic unique FString  (so it doesn't need to store it between calls) and the locking policy returning a unique token that then the caller does have to store. 

In the case of GAS it's easy to generate a deterministic unique FString, but I see Nic's point wondering if that will be true for all possible systems that may want to use the locking. 

Because these are going to be per-actor, and guaranteeing uniqueness across actors seems relatively easy, I'm inclined to leave the API the way it is. I think the normal case will be that unique and deterministic FStrings can be generated, so not having to store the locking policy generated token minimizes the footprint we have in the Unreal Engine itself.

I may also be a bit biased because I've always thought the debug string should be required, just to prevent people from getting into situations that are really difficult to debug.

Reply:
Ok, disagree and commit it is then.